### DISCUSSION TITLE
Solution: Nuxt app as Bun's single-file executable

### DISCUSSION THREAD

**Discussion by alpharder** [nuxt/nuxt#27746] [Sharing]
I needed to simplify my deployment processes so the whole application including static assets could be deployed as a single executable binary.

I've spent a couple hours tinkering with Nuxt, Nitro, Rollup and ESBuild documentations as well as reading Nitro source code in order to finally achieve Nuxt-based application properly working when complied to Bun's [Single-file Executable](https://bun.sh/docs/bundler/executables).

Sharing it here so it anyone trying to achieve the same could save some time. It could also be considered to be included to Nuxt's Bun-related documentation. 

`nuxt.config.ts`:

```ts
export default defineNuxtConfig({
  // ... rest of the config
  nitro: {
    preset: "bun",
    node: true,
    inlineDynamicImports: true,
    serveStatic: "inline",
    esbuild: {
      options: {
        target: "esnext",
      },
    },
  },
});
```

Then inside your project directory:
```sh
$ bun run build
$ bun build ./.output/server/index.mjs --compile --outfile ~/bun-built-binary
```

**Comment by ceferrari** [comment]
It did not work for me. I get the following errors on `bun build ...`:

```
1 | import { escapeHtml } from '@vue/shared';
                               ^
error: Could not resolve: "@vue/shared". Maybe you need to "bun install"?
    at /{ommited_path}/dist/server/chunks/_/error-500.mjs:1:28

1 | export * from '@vue/server-renderer'
                  ^
error: Could not resolve: "@vue/server-renderer". Maybe you need to "bun install"?
    at /{ommited_path}/dist/server/node_modules/vue/server-renderer/index.mjs:1:15

10 | var compilerDom = require('@vue/compiler-dom');
                               ^
error: Could not resolve: "@vue/compiler-dom". Maybe you need to "bun install"?
    at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:10:27

11 | var runtimeDom = require('@vue/runtime-dom');
                              ^
error: Could not resolve: "@vue/runtime-dom". Maybe you need to "bun install"?
    at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:11:26

12 | var shared = require('@vue/shared');
                          ^
error: Could not resolve: "@vue/shared". Maybe you need to "bun install"?
    at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:12:22
```

**Reply by th3l0g4n** [reply]
> It did not work for me. I get the following errors on `bun build ...`:
> 
> ```
> 1 | import { escapeHtml } from '@vue/shared';
>                                ^
> error: Could not resolve: "@vue/shared". Maybe you need to "bun install"?
>     at /{ommited_path}/dist/server/chunks/_/error-500.mjs:1:28
> 
> 1 | export * from '@vue/server-renderer'
>                   ^
> error: Could not resolve: "@vue/server-renderer". Maybe you need to "bun install"?
>     at /{ommited_path}/dist/server/node_modules/vue/server-renderer/index.mjs:1:15
> 
> 10 | var compilerDom = require('@vue/compiler-dom');
>                                ^
> error: Could not resolve: "@vue/compiler-dom". Maybe you need to "bun install"?
>     at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:10:27
> 
> 11 | var runtimeDom = require('@vue/runtime-dom');
>                               ^
> error: Could not resolve: "@vue/runtime-dom". Maybe you need to "bun install"?
>     at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:11:26
> 
> 12 | var shared = require('@vue/shared');
>                           ^
> error: Could not resolve: "@vue/shared". Maybe you need to "bun install"?
>     at /{ommited_path}/dist/server/node_modules/vue/dist/vue.cjs.js:12:22
> ```

Same here

**Reply by wling-art** [reply]
Same here

**Reply by linitsh** [reply]
the same - any solutions?

**Reply by BlackTeaLeaf** [reply]
After the Nuxt build process, the server/node_modules directory includes only production dependencies. Since bun build defaults process.env to development, it may attempt to import incorrect or missing versions. Therefore, the --production flag is required.

``` bash
bun build --compile --production ./.output/server/index.mjs --outfile server
```

**Comment by linitsh** [comment]
so i found solution for this issue:

```shell
bun run build
cd ./.output/server
rm -rf node_modules # required!!!
bun install
bun build --compile --minify --outfile ./nuxt ./index.mjs
```
result:
```shell
 [190ms]  minify  -0.86 MB (estimate)
   [8ms]  bundle  47 modules
 [277ms] compile  ./nuxt.exe
```

seems need change some builder options in nuxt.config to fix this - but im not a pro in nitro esbuild things :(

**Reply by olliepotter** [reply]
So I learned the reason this works... 

When you compile with bun, it has to resolve imports like this statically:
```
import { escapeHtml } from '@vue/shared';
```
But these imports **are not** bundled by Nuxt or Nitro. Instead, they're expected to be in `node_modules/` relative to `.output/server`, because Nuxt treats that folder like a deployable app.

Doing this:
```
cd .output/server
rm -rf node_modules
bun install
```
Creates a full standalone `node_modules` tree scoped to the output server — which Bun could then resolve during `--compile`.

So to get the same effect we need the build output of nitro to inline all the external dependencies, so we need to add the `noExternals` option to the `nuxt.config.ts`:
```ts
// https://nuxt.com/docs/api/configuration/nuxt-config
export default defineNuxtConfig({
  compatibilityDate: "2025-07-15",
  nitro: {
    preset: "bun",
    node: true,
    noExternals: true, <---------------- HERE
    inlineDynamicImports: true,
    serveStatic: "inline",
    esbuild: {
      options: {
        target: "esnext",
      },
    },
  },
  devtools: { enabled: true },
});
```

Then finally:
```
nuxt build
bun build .output/server/index.mjs --compile --outfile nuxtbin
```

**Comment by Hugo-Dz** [comment]
Played around this issue, I managed to get something working but still in progress: https://github.com/Hugo-Dz/exe/tree/main/packages/nuxt

**Comment by wodepig** [comment]
I bypassed the packaging problem, but took the. output directory as a compressed file, downloaded it after the program started, and then used node index.mjs to start it
python startup: https://github.com/wodepig/node-startup-exe
electron startup: https://github.com/wodepig/xxdl-electron

**Comment by jprando** [comment]
🙋‍♂️ Hi everyone,

I’ve been following this discussion closely as I share a great interest in deep Bun integration within the Nuxt ecosystem.

To address a specific challenge: simplifying the process of generating a standalone binary from a Nuxt project using Bun

I’ve implemented and published a new module: [nuxt-bun-compile](https://github.com/jprando/nuxt-bun-compile).

My goal with this implementation was to bridge the gap between Nuxt's build output and Bun's native compilation capabilities. The module allows you to:

Easily wrap your Nuxt application into a single executable binary.

Leverage `Bun.build` and `bun build --compile` seamlessly within the Nuxt workflow.

Streamline deployments by providing a portable, high-performance output.

If you are looking for an efficient way to compile your Nuxt apps into binaries today, I’d love for you to give it a try. Any feedback or contributions to make the Bun experience even more elegant for the community are very welcome!

Best regards, [@jprando](https://github.com/jprando)

> A module for Bun's native compilation in Nuxt
